File: src/Agent/Agent.Core/Models/Api/v1/McpToolExecution.cs
// ------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All rights reserved.
// ------------------------------------------------------------

using System.Text.Json.Serialization;

namespace Agent.Core.Models.Api.v1;

/// <summary>
/// Represents an MCP (Model Context Protocol) tool execution with its parameters and status.
/// </summary>
public class McpToolExecution
{
    /// <summary>
    /// Unique identifier for this execution.
    /// </summary>
    public Guid Id { get; set; } = Guid.NewGuid();

    /// <summary>
    /// The full tool name (e.g., "kusto-mcp_kusto_query").
    /// </summary>
    public string FullToolName { get; set; } = string.Empty;

    /// <summary>
    /// The MCP server name (e.g., "kusto-mcp").
    /// </summary>
    public string McpServerName { get; set; } = string.Empty;

    /// <summary>
    /// The tool name without the MCP prefix (e.g., "kusto_query").
    /// </summary>
    public string ToolName { get; set; } = string.Empty;

    /// <summary>
    /// User-friendly display name for the tool (e.g., "Kusto Query").
    /// </summary>
    public string DisplayName { get; set; } = string.Empty;

    /// <summary>
    /// Current execution status.
    /// </summary>
    public McpToolExecutionStatus Status { get; set; } = McpToolExecutionStatus.Running;

    /// <summary>
    /// When the execution started.
    /// </summary>
    public DateTime StartedAt { get; set; } = DateTime.UtcNow;

    /// <summary>
    /// When the execution completed (if finished).
    /// </summary>
    public DateTime? CompletedAt { get; set; }

    /// <summary>
    /// Structured parameters for the tool execution.
    /// </summary>
    public McpToolParameters? Parameters { get; set; }

    /// <summary>
    /// The result of the execution (if completed).
    /// </summary>
    public string? Result { get; set; }

    /// <summary>
    /// Error message if the execution failed.
    /// </summary>
    public string? Error { get; set; }
}

/// <summary>
/// Status of an MCP tool execution.
/// </summary>
public enum McpToolExecutionStatus
{
    /// <summary>
    /// Tool is currently executing.
    /// </summary>
    Running,

    /// <summary>
    /// Tool completed successfully.
    /// </summary>
    Completed,

    /// <summary>
    /// Tool execution failed.
    /// </summary>
    Failed,

    /// <summary>
    /// Tool execution was cancelled.
    /// </summary>
    Cancelled
}

/// <summary>
/// Parameters for an MCP tool execution, structured by tool type.
/// </summary>
public class McpToolParameters
{
    // Common parameters
    /// <summary>
    /// Raw parameters dictionary for any tool.
    /// </summary>
    public Dictionary<string, object?>? Raw { get; set; }

    // Kusto-specific parameters
    /// <summary>
    /// Kusto cluster URL (for Kusto tools).
    /// </summary>
    public string? ClusterUrl { get; set; }

    /// <summary>
    /// Kusto database name (for Kusto tools).
    /// </summary>
    public string? Database { get; set; }

    /// <summary>
    /// KQL query (for kusto_query tool).
    /// </summary>
    public string? Query { get; set; }

    /// <summary>
    /// Table name (for kusto_sample, kusto_table_schema tools).
    /// </summary>
    public string? TableName { get; set; }

    /// <summary>
    /// Sample size (for kusto_sample tool).
    /// </summary>
    public int? SampleSize { get; set; }

    // ADO-specific parameters
    /// <summary>
    /// Repository name (for ADO tools).
    /// </summary>
    public string? Repository { get; set; }

    /// <summary>
    /// Source branch (for PR creation).
    /// </summary>
    public string? SourceBranch { get; set; }

    /// <summary>
    /// Target branch (for PR creation).
    /// </summary>
    public string? TargetBranch { get; set; }

    /// <summary>
    /// Pull request title.
    /// </summary>
    public string? Title { get; set; }

    /// <summary>
    /// Pull request description.
    /// </summary>
    public string? Description { get; set; }
}

File: src/Agent/Agent.Web/Client/src/src/Common/Contracts/DataPlane/McpToolExecution.ts
/**
 * Represents an MCP (Model Context Protocol) tool execution with its parameters and status.
 */
export interface McpToolExecution {
    /** Unique identifier for this execution */
    id: string;
    /** The full tool name (e.g., "kusto-mcp_kusto_query") */
    fullToolName: string;
    /** The MCP server name (e.g., "kusto-mcp") */
    mcpServerName: string;
    /** The tool name without the MCP prefix (e.g., "kusto_query") */
    toolName: string;
    /** User-friendly display name for the tool (e.g., "Kusto Query") */
    displayName: string;
    /** Current execution status */
    status: McpToolExecutionStatus;
    /** When the execution started */
    startedAt: string;
    /** When the execution completed (if finished) */
    completedAt?: string;
    /** Structured parameters for the tool execution */
    parameters?: McpToolParameters;
    /** The result of the execution (if completed) */
    result?: string;
    /** Error message if the execution failed */
    error?: string;
}

/**
 * Status of an MCP tool execution.
 */
export type McpToolExecutionStatus = 'Running' | 'Completed' | 'Failed' | 'Cancelled';

/**
 * Parameters for an MCP tool execution, structured by tool type.
 */
export interface McpToolParameters {
    /** Raw parameters dictionary for any tool */
    raw?: Record<string, unknown>;

    // Kusto-specific parameters
    /** Kusto cluster URL (for Kusto tools) */
    clusterUrl?: string;
    /** Kusto database name (for Kusto tools) */
    database?: string;
    /** KQL query (for kusto_query tool) */
    query?: string;
    /** Table name (for kusto_sample, kusto_table_schema tools) */
    tableName?: string;
    /** Sample size (for kusto_sample tool) */
    sampleSize?: number;

    // ADO-specific parameters
    /** Repository name (for ADO tools) */
    repository?: string;
    /** Source branch (for PR creation) */
    sourceBranch?: string;
    /** Target branch (for PR creation) */
    targetBranch?: string;
    /** Pull request title */
    title?: string;
    /** Pull request description */
    description?: string;
}

/**
 * Configuration for how to display MCP tools in the UI.
 */
export interface McpToolDisplayConfig {
    /** Icon name for the tool type */
    icon: string;
    /** Color scheme for the card */
    colorScheme: 'kusto' | 'ado' | 'generic';
    /** Whether to show a code editor for query-type parameters */
    showQueryEditor: boolean;
    /** Language for syntax highlighting (e.g., 'kql', 'sql') */
    queryLanguage?: string;
}

/**
 * Gets the display configuration for an MCP tool based on its name.
 */
export const getMcpToolDisplayConfig = (mcpServerName: string, toolName: string): McpToolDisplayConfig => {
    // Kusto MCP tools
    if (mcpServerName === 'kusto-mcp') {
        return {
            icon: 'Database',
            colorScheme: 'kusto',
            showQueryEditor: toolName === 'kusto_query',
            queryLanguage: 'kql',
        };
    }

    // Azure DevOps MCP tools
    if (mcpServerName === 'ado-mcp') {
        return {
            icon: 'GitPullRequest',
            colorScheme: 'ado',
            showQueryEditor: false,
        };
    }

    // Default/generic MCP tools
    return {
        icon: 'Puzzle',
        colorScheme: 'generic',
        showQueryEditor: false,
    };
};

/**
 * Parses an MCP tool name into its components.
 * Format: "{mcp-server-name}_{tool_name}" (e.g., "kusto-mcp_kusto_query")
 */
export const parseMcpToolName = (fullToolName: string): { mcpServerName: string; toolName: string; displayName: string } => {
    const firstUnderscoreIndex = fullToolName.indexOf('_');

    if (firstUnderscoreIndex === -1) {
        return {
            mcpServerName: fullToolName,
            toolName: fullToolName,
            displayName: formatDisplayName(fullToolName),
        };
    }

    const mcpServerName = fullToolName.substring(0, firstUnderscoreIndex);
    const toolName = fullToolName.substring(firstUnderscoreIndex + 1);

    return {
        mcpServerName,
        toolName,
        displayName: formatDisplayName(toolName),
    };
};

/**
 * Formats a tool name into a user-friendly display name.
 * e.g., "kusto_query" -> "Kusto Query"
 */
const formatDisplayName = (toolName: string): string => {
    return toolName
        .replace(/_/g, ' ')
        .replace(/-/g, ' ')
        .split(' ')
        .map(word => word.charAt(0).toUpperCase() + word.slice(1).toLowerCase())
        .join(' ');
};
